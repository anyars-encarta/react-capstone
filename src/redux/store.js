// store.js
import { configureStore } from '@reduxjs/toolkit';
import continentsReducer from './continentsSlice';
import countriesReducer from './countriesSlice';

const store = configureStore({
  reducer: {
    continents: continentsReducer,
    countries: countriesReducer,
  },
});

export default store;
