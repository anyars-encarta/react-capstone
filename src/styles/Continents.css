/* Continents.css */
@import url('https://fonts.googleapis.com/css2?family=Roboto+Slab:wght@100;200;300;400;500;600;700;800;900&display=swap');

* {
  font-family: 'Roboto Slab', serif;
}

.top-stats {
  display: flex;
  flex-direction: column;
  height: 200px;
  padding: 10px;
  margin-top: 70px;
  background-image: linear-gradient(100deg, rgba(0, 0, 0, 1.75), rgba(225, 73, 132, 0.75)), url('../assets/coronavirus-ikages.jpg');
  background-repeat: no-repeat;
  background-size: cover;
  color: white;
  justify-content: space-between;
}

.count {
  margin: 0;
  text-align: right;
}

.top {
  text-align: right;
}

.top-stats .global {
  margin: 0;
  font-size: 40px;
}

.top-stats .count {
  font-size: 15px;
}

.top-stats .arrow-right {
  align-self: flex-end;
}

.continent-stats {
  margin: 0;
  padding: 2px;
  background-color: #e14984;
  color: #ffff;
  font-weight: bold;
}

.continent-container {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
  background-color: rgba(225, 73, 132, 0.75);
}

.continent {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  justify-content: right;
  width: 50%;
  padding: 10px;
  color: white;
  background-size: cover;
  position: relative;
  cursor: pointer;
  margin: 0;
  box-sizing: border-box;
  background-color: rgba(0, 0, 0, 0.5);
}

.overlay {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5); /* Adjust the alpha value (last parameter) for opacity */
  z-index: 1; /* Make sure the overlay is above the background image */
}

.continent:nth-child(4n-1),
.continent:nth-child(4n-2) {
  background-color: rgba(225, 73, 132); /* Background color for cards 3 and 4 */
}

.info {
  display: flex;
  flex-direction: column;
  z-index: 99;
  padding: 40px 0 40px 0;
}

.view-more {
  z-index: 99;
}

.info p {
  text-align: right;
  margin: 0;
}
