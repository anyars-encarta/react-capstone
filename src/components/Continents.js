// Continents.js
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom'; // Import useHistory
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCircleRight } from '@fortawesome/free-solid-svg-icons';
import { setContinents, selectContinents } from '../redux/continentsSlice';
import Countries from './Countries';
import ContinentMap from './ContinentMap';
import '../styles/Continents.css';

function Continents() {
  const dispatch = useDispatch();
  const continents = useSelector(selectContinents);
  const [selectedContinent, setSelectedContinent] = useState(null);
  const [showCountries, setShowCountries] = useState(false);

  const history = useHistory(); // Use the useHistory hook

  let totalCases = 0;
  let totalRecoveries = 0;
  let totalDeaths = 0;

  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await fetch('https://corona.lmao.ninja/v2/continents');
        const data = await response.json();
        dispatch(setContinents(data)); // Dispatch the action to set continents
      } catch (error) {
        throw new Error('Error fetching data:', error);
      }
    };

    fetchData();
  }, [dispatch]);

  if (Array.isArray(continents)) {
    continents.forEach((continent) => {
      totalCases += parseInt(continent.cases, 10);
      totalRecoveries += parseInt(continent.recovered, 10);
      totalDeaths += parseInt(continent.deaths, 10);
    });
  }

  const formatNumber = (number) => number.toLocaleString();

  const handleContinentClick = (continentName) => {
    setSelectedContinent(continentName);
    setShowCountries(true);
    history.push(`/countries/${continentName}`);
  };

  return (
    <>
      <div className="top-stats">
        <div className="top">
          <p className="global">Global Stats</p>
          <p className="count">
            Cases:
            {' '}
            {totalCases.toLocaleString()}
          </p>
          <p className="count">
            Recoveries:
            {' '}
            {totalRecoveries.toLocaleString()}
          </p>
          <p className="count">
            Deaths:
            {' '}
            {totalDeaths.toLocaleString()}
          </p>
        </div>
        <FontAwesomeIcon icon={faCircleRight} className="arrow-right" />
      </div>
      <p className="continent-stats">CONTINENT STATS</p>
      <div className="continent-container">
        {continents.map((continent) => (
          <div
            key={continent.continent}
            className="continent"
            style={{ backgroundImage: `url(${ContinentMap[continent.continent]})` }}
            onClick={() => handleContinentClick(continent.continent)}
            role="button" // Add role="button" for accessibility
            tabIndex="0" // Add tabIndex to make it focusable
            onKeyDown={(e) => {
              if (e.key === 'Enter') {
                // Handle Enter key press
                handleContinentClick(continent.continent);
              }
            }}
          >
            <div className="overlay" />
            <div className="info">
              <p className="continent-name">{continent.continent}</p>
              <p className="cases-count">
                Cases:
                {' '}
                {formatNumber(continent.cases)}
              </p>
            </div>
            <FontAwesomeIcon icon={faCircleRight} className="view-more" />
          </div>
        ))}
      </div>
      {showCountries && <Countries continentName={selectedContinent} />}
    </>
  );
}

export default Continents;
