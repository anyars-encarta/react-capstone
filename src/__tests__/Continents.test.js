import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import store from '../redux/store';
import Continents from '../components/Continents';

describe('Continents component', () => {
  it('should render correctly', async () => {
    render(
      <Provider store={store}>
        <Continents />
      </Provider>,
    );

    await waitFor(() => {
      const { continents } = store.getState();
      return continents && continents.length > 0;
    });

    expect(screen.getByText('Global Stats')).toBeInTheDocument();
    expect(screen.getByText('CONTINENT STATS')).toBeInTheDocument();

    const { continents } = store.getState();
    if (continents.length > 0) {
      const continent = continents[0];
      const continentElement = screen.getByText(continent.continent);
      expect(continentElement).toBeInTheDocument();
    }
  });

  it('should navigate to countries page when a continent is clicked', async () => {
    const history = createMemoryHistory();
    render(
      <Provider store={store}>
        <Router history={history}>
          <Continents />
        </Router>
      </Provider>,
    );

    await waitFor(() => screen.getByText('Global Stats'));

    const { continents } = store.getState();
    if (continents.length > 0) {
      const continent = continents[0];
      const continentElement = screen.getByText(continent.continent);
      continentElement.click();
      expect(history.location.pathname).toBe(`/countries/${continent.continent}`);
    }
  });

  describe('Continents component', () => {
    it('should match snapshot', async () => {
      const { container } = render(
        <Provider store={store}>
          <Continents />
        </Provider>,
      );

      await waitFor(() => screen.getByText('Global Stats'));
      expect(container).toMatchSnapshot();
    });
  });
});
