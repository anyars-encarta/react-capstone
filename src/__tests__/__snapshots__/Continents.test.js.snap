// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Continents component Continents component should match snapshot 1`] = `
<div>
  <div
    class="top-stats"
  >
    <div
      class="top"
    >
      <p
        class="global"
      >
        Global Stats
      </p>
      <p
        class="count"
      >
        Cases:
         
        0
      </p>
      <p
        class="count"
      >
        Recoveries:
         
        0
      </p>
      <p
        class="count"
      >
        Deaths:
         
        0
      </p>
    </div>
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-circle-right arrow-right"
      data-icon="circle-right"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 256a256 256 0 1 0 512 0A256 256 0 1 0 0 256zM294.6 135.1l99.9 107.1c3.5 3.8 5.5 8.7 5.5 13.8s-2 10.1-5.5 13.8L294.6 376.9c-4.2 4.5-10.1 7.1-16.3 7.1C266 384 256 374 256 361.7l0-57.7-96 0c-17.7 0-32-14.3-32-32l0-32c0-17.7 14.3-32 32-32l96 0 0-57.7c0-12.3 10-22.3 22.3-22.3c6.2 0 12.1 2.6 16.3 7.1z"
        fill="currentColor"
      />
    </svg>
  </div>
  <p
    class="continent-stats"
  >
    CONTINENT STATS
  </p>
  <div
    class="continent-container"
  />
</div>
`;
